README.md
# wow-gear-stats-exporter
Export your WoW characters gear and stats to CSV format


CHANGELOG.md
**1.0**

- Initial release.


Modules/DB.lua
----------
-- Corruption database shamelessly stolen from CorruptionTooltips - https://www.curseforge.com/wow/addons/corruption-tooltips
-- Credits: suspectz (Anayanka - Defias Brotherhood - EU)
----------

local Addon = LibStub("AceAddon-3.0"):GetAddon("GearStatsExporter")
local Module = Addon:NewModule("DB")

local bonuses = {
   ["6483"] = {"Avoidant", "I", 315607},
   ["6484"] = {"Avoidant", "II", 315608},
   ["6485"] = {"Avoidant", "III", 315609},
   ["6474"] = {"Expedient", "I", 315544},
   ["6475"] = {"Expedient", "II", 315545},
   ["6476"] = {"Expedient", "III", 315546},
   ["6471"] = {"Masterful", "I", 315529},
   ["6472"] = {"Masterful", "II", 315530},
   ["6473"] = {"Masterful", "III", 315531},
   ["6480"] = {"Severe", "I", 315554},
   ["6481"] = {"Severe", "II", 315557},
   ["6482"] = {"Severe", "III", 315558},
   ["6477"] = {"Versatile", "I", 315549},
   ["6478"] = {"Versatile", "II", 315552},
   ["6479"] = {"Versatile", "III", 315553},
   ["6493"] = {"Siphoner", "I", 315590},
   ["6494"] = {"Siphoner", "II", 315591},
   ["6495"] = {"Siphoner", "III", 315592},
   ["6437"] = {"Strikethrough", "I", 315277},
   ["6438"] = {"Strikethrough", "II", 315281},
   ["6439"] = {"Strikethrough", "III", 315282},
   ["6555"] = {"Racing Pulse", "I", 318266},
   ["6559"] = {"Racing Pulse", "II", 318492},
   ["6560"] = {"Racing Pulse", "III", 318496},
   ["6556"] = {"Deadly Momentum", "I", 318268},
   ["6561"] = {"Deadly Momentum", "II", 318493},
   ["6562"] = {"Deadly Momentum", "III", 318497},
   ["6558"] = {"Surging Vitality", "I", 318270},
   ["6565"] = {"Surging Vitality", "II", 318495},
   ["6566"] = {"Surging Vitality", "III", 318499},
   ["6557"] = {"Honed Mind", "I", 318269},
   ["6563"] = {"Honed Mind", "II", 318494},
   ["6564"] = {"Honed Mind", "III", 318498},
   ["6549"] = {"Echoing Void", "I", 318280},
   ["6550"] = {"Echoing Void", "II", 318485},
   ["6551"] = {"Echoing Void", "III", 318486},
   ["6552"] = {"Infinite Stars", "I", 318274},
   ["6553"] = {"Infinite Stars", "II", 318487},
   ["6554"] = {"Infinite Stars", "III", 318488},
   ["6547"] = {"Ineffable Truth", "I", 318303},
   ["6548"] = {"Ineffable Truth", "II", 318484},
   ["6537"] = {"Twilight Devastation", "I", 318276},
   ["6538"] = {"Twilight Devastation", "II", 318477},
   ["6539"] = {"Twilight Devastation", "III", 318478},
   ["6543"] = {"Twisted Appendage", "I", 318481},
   ["6544"] = {"Twisted Appendage", "II", 318482},
   ["6545"] = {"Twisted Appendage", "III", 318483},
   ["6540"] = {"Void Ritual", "I", 318286},
   ["6541"] = {"Void Ritual", "II", 318479},
   ["6542"] = {"Void Ritual", "III", 318480},
   ["6573"] = {"Gushing Wound", "", 318272},
   ["6546"] = {"Glimpse of Clarity", "", 318239},
   ["6571"] = {"Searing Flames", "", 318293},
   ["6572"] = {"Obsidian Skin", "", 316651},
   ["6567"] = {"Devour Vitality", "", 318294},
   ["6568"] = {"Whispered Truths", "", 316780},
   ["6570"] = {"Flash of Insight", "", 318299},
   ["6569"] = {"Lash of the Void", "", 317290}
}

local gemStats = {
   ["empty"] = {
      empty = true
   },
   ["168642"] = {
      name = "Versatile Dark Opal",
      stat = "Versatility",
      value = 50
   },
   ["168640"] = {
      name = "Masterful Sea Currant",
      stat = "Mastery",
      value = 50
   },
   ["168639"] = {
      name = "Deadly Lava Lazuli",
      stat = "Critical Strike",
      value = 50
   },
   ["168641"] = {
      name = "Quick Sand Spinel",
      stat = "Haste",
      value = 50
   },
   ["168637"] = {
      name = "Leviathan's Eye of Agility",
      stat = "Agility",
      value = 120
   },
   ["168638"] = {
      name = "Leviathan's Eye of Intellect",
      stat = "Intellect",
      value = 120
   },
   ["168636"] = {
      name = "Leviathan's Eye of Strength",
      stat = "Strength",
      value = 120
   }
}

local slotNames = {
   [1] = "Head",
   [2] = "Neck",
   [3] = "Shoulder",
   [4] = "Shirt",
   [5] = "Chest",
   [6] = "Waist",
   [7] = "Legs",
   [8] = "Feet",
   [9] = "Wrist",
   [10] = "Hands",
   [11] = "Finger 1",
   [12] = "Finger 2",
   [13] = "Trinket 1",
   [14] = "Trinket 2",
   [15] = "Back",
   [16] = "Main Hand",
   [17] = "Off Hand",
   [18] = "Tabard",
   [19] = "Ammo"
}

local modifiers = {
   [1] = "Armor",
   [2] = "Stamina",
   [3] = "Agility",
   [4] = "Strength",
   [5] = "Intellect",
   [6] = "Corruption",
   [7] = "Critical Strike",
   [8] = "Haste",
   [9] = "Mastery",
   [10] = "Versatility",
   [11] = "Socket",
   [12] = "Speed",
   [13] = "Leech",
   [14] = "Avoidance",
   [15] = "Indestructible"
}

local extraModifiers = {
   [1] = "Speed",
   [2] = "Leech",
   [3] = "Avoidance",
   [4] = "Indestructible"
}

local socketModifiers = {
   [1] = "Agility",
   [2] = "Strength",
   [3] = "Intelligence",
   [4] = "Critical Strike",
   [5] = "Haste",
   [6] = "Mastery",
   [7] = "Versatility"
}

local headerOrder = {
   ["Armor"] = false,
   ["Stamina"] = false,
   ["Agility"] = false,
   ["Strength"] = false,
   ["Intellect"] = false,
   ["Corruption"] = false,
   ["Critical Strike"] = false,
   ["Haste"] = false,
   ["Mastery"] = false,
   ["Versatility"] = false,
   ["Socket"] = false,
   ["Speed"] = false,
   ["Leech"] = false,
   ["Avoidance"] = false,
   ["Indestructible"] = false
}

function Module:GetBonus(bonusID)
   return bonuses[tostring(bonusID)]
end

function Module:GetGemStats(gemID)
   return gemStats[tostring(gemID)]
end

function Module:GetSlotName(slotId)
   return slotNames[slotId]
end

function Module:GetSlots()
   return slotNames
end

function Module:GetModifiers()
   return modifiers
end

function Module:GetExtraModifiers()
   return extraModifiers
end

function Module:GetSocketModifiers()
   return socketModifiers
end

function Module:GetHeaderOrder()
   return headerOrder
end

-- Get item modifier name from global constants
function Module:GetItemModifierName(constant)
   return _G[constant]
end

-- Get equipment slot name from global constants
function Module:GetItemEquipLocationName(itemEquipLoc)
   return slotNames[itemEquipLoc]
end


Modules/EventHandler.lua
local Addon = LibStub("AceAddon-3.0"):GetAddon("GearStatsExporter")
local Module = Addon:NewModule("EventHandler")
local Scanner

local frame, events = CreateFrame("Frame"), {};

-- Called when INSPECT_READY event is fired
function events:INSPECT_READY(...)
    local guid = ...
    Scanner:ScanPlayerGear()
 end

function Module:OnInitialize()

    Scanner = Addon:GetModule("Scanner")
        -- Setup event catching
    frame:SetScript("OnEvent", function(self, event, ...)
        events[event](self, ...);
    end);

    for k, v in pairs(events) do
        frame:RegisterEvent(k);
    end
end

function Module:NotifyScan()
    NotifyInspect("player")
end

SLASH_GEAR_STATS_EXPORTER1 = '/gse'
function SlashCmdList.GEAR_STATS_EXPORTER(msg, editbox)
    Module:NotifyScan()
end


Modules/Scanner.lua
local Addon = LibStub("AceAddon-3.0"):GetAddon("GearStatsExporter")
local Module = Addon:NewModule("Scanner")

local itemTable = {
    ["Head"] = nil,
    ["Neck"] = nil,
    ["Shoulder"] = nil,
    ["Back"] = nil,
    ["Chest"] = nil,
    ["Wrist"] = nil,
    ["Hands"] = nil,
    ["Waist"] = nil,
    ["Legs"] = nil,
    ["Feet"] = nil,
    ["Finger 1"] = nil,
    ["Finger 2"] = nil,
    ["Trinket 1"] = nil,
    ["Trinket 2"] = nil,
    ["Main Hand"] = nil,
    ["Off Hand"] = nil
}

local DB, EventHandler

function Module:OnInitialize()
    DB = Addon:GetModule("DB")
    EventHandler = Addon:GetModule("EventHandler")
end

function printTable(t, tab)
    tab = tab .. "   "
    for k, v in pairs(t) do
        if type(v) == "table" then
            print(tab, k .. ":")
            printTable(v, tab)
        else
            print(tab, k .. ": " .. tostring(v))
        end
    end
end

function printTable2(t)
    for i, v in ipairs(DB:GetSlots()) do
        print(i, v)
    end
end

function appendToString(str1, str2)
    return str1 .. "," .. str2
end

function Module:ScanPlayerGear()
    local slots = DB:GetSlots()

    for equipmentSlotIndex, equipmentSlotName in ipairs(slots) do
        -- print(equipmentSlotIndex, equipmentSlotName)
        local itemLocation = ItemLocation:CreateFromEquipmentSlot(equipmentSlotIndex)
        local item = Item:CreateFromItemLocation(itemLocation)

        if itemLocation:IsValid() and item:IsItemEmpty() == false then
            local inventoryItemLink = GetInventoryItemLink("player", equipmentSlotIndex)

            self:HandleItem(item, inventoryItemLink, equipmentSlotName)
        end
    end

    self:MakeCsv()
end

function Module:MakeCsv()
    local slots = DB:GetSlots()

    local playerName, realm = UnitName("player")
    local header = self:MakeHeader()

    -- print(header)

    for _, equipmentSlotName in ipairs(slots) do
        local item = itemTable[equipmentSlotName]

        if item ~= nil then
            local line = ""
            line = line .. "," .. equipmentSlotName
            line = line .. "," .. item.name
            line = line .. "," .. item.itemLevel

            for _, modifier in ipairs(DB:GetModifiers()) do
                if item.itemStats[modifier] then
                    line = line .. "," .. item.itemStats[modifier]
                end
            end

        -- for _, modifier in ipairs(DB:GetExtraModifiers()) do
        --     if item.itemStats[modifier] then
        --         line = line .. "," .. item.itemStats[modifier]
        --     end
        -- end
        end
    end
end

function Module:MakeHeader()
    local slots = DB:GetSlots()
    local hasCorruption = false
    local socketModifiers = {}
    local modifiersOnGear = {
        ["Armor"] = false,
        ["Stamina"] = false,
        ["Agility"] = false,
        ["Strength"] = false,
        ["Intellect"] = false,
        ["Corruption"] = false,
        ["Critical Strike"] = false,
        ["Haste"] = false,
        ["Mastery"] = false,
        ["Versatility"] = false,
        ["Socket"] = false,
        ["Speed"] = false,
        ["Leech"] = false,
        ["Avoidance"] = false,
        ["Indestructible"] = false
    }

    for _, equipmentSlotName in ipairs(slots) do
        local item = itemTable[equipmentSlotName]

        if item ~= nil then
            for modifier, _ in pairs(item.itemStats) do
                modifiersOnGear[modifier] = true

                if modifier == "Socket" then
                    local socket = item.itemStats[modifier]
                    socketModifiers[socket.stat] = true
                end
            end
        end
    end

    local header = "Slot,Name,iLvl"

    for _, modifier in ipairs(DB:GetModifiers()) do
        if modifiersOnGear[modifier] then
            if modifier == "Socket" then
                for _, socketModifier in ipairs(DB:GetSocketModifiers()) do
                    if socketModifiers[socketModifier] then
                        header = header .. ", Socket " .. socketModifier
                    end
                end
            elseif modifier == "Corruption" and hasCorruption then
                header = header .. ",Corruption,Corruption Value"
            else
                header = header .. "," .. modifier
            end
        end
    end

    return header
end

-- Process item and create a table with all relevant stats
function Module:HandleItem(item, inventoryItemLink, equipmentSlotName)
    local itemName,
        itemLink,
        itemRarity,
        itemLevel,
        itemMinLevel,
        itemType,
        itemSubType,
        itemStackCount,
        itemEquipLoc,
        itemTexture,
        itemSellPrice = GetItemInfo(inventoryItemLink)

    -- print(itemName, itemLink, itemRarity, itemLevel, itemMinLevel, itemType, itemSubType, itemStackCount, itemEquipLoc, itemTexture, itemSellPrice)

    itemTable[equipmentSlotName] = {
        name = itemName,
        itemLevel = itemLevel,
        itemStats = self:CreateItemStats(inventoryItemLink)
    }

    -- printTable(itemTable[equipmentSlotName], "")
end

-- Create custom item stats
function Module:CreateItemStats(itemLink)
    local customItemStats = {}
    local itemStats = GetItemStats(itemLink)
    local corruption = self:GetCorruptionByItemLink(itemLink)
    local corruptionValue = nil

    for key, value in pairs(itemStats) do
        local modName = DB:GetItemModifierName(key)

        print(key)

        if key == "EMPTY_SOCKET_PRISMATIC" then
            customItemStats["Socket"] = self:GetGemStats(itemStats, itemLink)
        elseif key == "ITEM_MOD_CORRUPTION" then
            corruptionValue = value
        else
            -- print(modName)
            customItemStats[modName] = value
        end
    end

    if corruption ~= nil then
        corruption.value = corruptionValue
        customItemStats["Corruption"] = corruption
    end

    return customItemStats
end

-- Create custom gem stats
function Module:GetGemStats(itemStats, itemLink)
    local gemName, gemItemLink = GetItemGem(itemLink, 1)

    if gemName == nil then
        return DB:GetGemStats("empty")
    end

    local gemItem = Item:CreateFromItemLink(gemItemLink)
    local gemItemID = gemItem:GetItemID()

    if DB:GetGemStats(gemItemID) == nil then
        return DB:GetGemStats("empty")
    end

    return DB:GetGemStats(gemItemID)
end

function Module:GetCorruptionByID(bonusID)
    local corruption = DB:GetBonus(bonusID)

    if corruption ~= nil then
        return {
            name = corruption[1],
            rank = corruption[2]
        }
    end
end

function Module:GetItemSplit(itemLink)
    local itemString = string.match(itemLink, "item:([%-?%d:]+)")
    local itemSplit = {}

    if itemString ~= nil then
        -- Split data into a table
        for _, v in ipairs({strsplit(":", itemString)}) do
            if v == "" then
                itemSplit[#itemSplit + 1] = 0
            else
                itemSplit[#itemSplit + 1] = tonumber(v)
            end
        end
    end

    return itemSplit
end

function Module:GetCorruptionByItemLink(itemLink)
    local itemSplit = self:GetItemSplit(itemLink)
    local bonuses = {}

    if IsCorruptedItem(itemLink) then
        for index = 1, itemSplit[13] do
            bonuses[#bonuses + 1] = itemSplit[13 + index]
        end
    end

    if #bonuses > 0 then
        for _, bonusID in pairs(bonuses) do
            local corruption = self:GetCorruptionByID(bonusID)
            if corruption ~= nil then
                return corruption
            end
        end
    end
end
